### PR TITLE
Several caching improvements:

### PR DESCRIPTION
**Pull request by mystenmark**
- Populate the cache with the latest object version when possible.
- Add negative caching for getting object by id.
- Use latest version cache for satisfying dynamic field reads

### CODE REVIEW

File: crates/sui-core/src/execution_cache/writeback_cache.rs [hunk 1]
```diff
@@ -1060,9 +1153,7 @@ impl ObjectCacheRead for WritebackCache {
                 ObjectEntry::Deleted => (object_id, version, ObjectDigest::OBJECT_DIGEST_DELETED),
```
Comment (halfprice) [inline crates/sui-core/src/execution_cache/writeback_cache.rs:1153]: Similar to above comment, can we still get `ObjectEntry::Deleted` or ObjectEntry::Wrapped` here?

**Review by halfprice — APPROVED** [review]
LGTM with one comment about whether we are still able to read Deleted or Wrapped objects.

File: crates/sui-core/src/execution_cache/writeback_cache.rs [hunk 2]
```diff
@@ -485,6 +508,17 @@ impl WritebackCache {
         &self,
         object_id: &ObjectID,
     ) -> CacheResult<(SequenceNumber, ObjectEntry)> {
+        if let Some(entry) = self.cached.object_by_id_cache.get(object_id) {
+            let entry = entry.lock();
+            let (latest_version, latest_object) = &*entry;
+            if latest_object.is_live() {
```
Comment (halfprice) [inline crates/sui-core/src/execution_cache/writeback_cache.rs:514]: Ok, reading the PR again, I think I understand what `is_live` means, which basically try to accomodate `cache_object_not_found`. I think the criteria should be `latest_version == 0 && entry == ObjectEntry::Deleted`, right?

Or can we create an ObjectEntry::NotExist?

**Review by halfprice — APPROVED** [review]

File: crates/sui-core/src/execution_cache/writeback_cache.rs [hunk 3]
```diff
@@ -865,6 +897,61 @@ impl WritebackCache {
...
+            // only one racing insert will call the closure
+            .or_insert_with(|| Arc::new(Mutex::new((version, object.clone()))));
+
+        // We may be racing with another thread that observed an older version of the object
+        if !entry.is_fresh() {
+            // !is_fresh means we lost the race, and actual_entry holds the entry that was
+            // inserted by the other thread. We need to check if we have a more recent version
+            // than the other reader.
+            // This could also mean that the entry was inserted by a transaction write. This
+            // could occur in the following case:
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-core/src/execution_cache/writeback_cache.rs:960]: `is_fresh` means that our init callback ran, so `!is_fresh` means that we need to check if we have a more recent version that what is in the cash. I will try to clarify the comment

File: crates/sui-core/src/execution_cache/writeback_cache.rs (hunk 1 above)
Comment (mystenmark) [reply crates/sui-core/src/execution_cache/writeback_cache.rs:1184]: Cleaned this up by having a different enum type for the latest-object cache

File: crates/sui-core/src/execution_cache/writeback_cache.rs [hunk 4]
```diff
@@ -865,6 +897,61 @@ impl WritebackCache {
...
+            //
+            // THREAD 1            | THREAD 2
+            // reads object at v1  |
+            //                     | tx writes object at v2
+            // tries to cache v1
+            //
+            // Thread 1 will see that v2 is already in the cache
+
+            // this point because there should have been a cache hit.)
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-core/src/execution_cache/writeback_cache.rs:969]: sorry, just an unfinished comment

File: crates/sui-core/src/execution_cache/writeback_cache.rs [hunk 5]
```diff
@@ -485,6 +508,17 @@ impl WritebackCache {
         &self,
         object_id: &ObjectID,
     ) -> CacheResult<(SequenceNumber, ObjectEntry)> {
+        if let Some(entry) = self.cached.object_by_id_cache.get(object_id) {
+            let entry = entry.lock();
+            let (latest_version, latest_object) = &*entry;
+            if latest_object.is_live() {
+                return CacheResult::Hit((*latest_version, latest_object.clone()));
+            } else {
+                assert_eq!(*latest_version, SequenceNumber::from_u64(0));
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply crates/sui-core/src/execution_cache/writeback_cache.rs]: cleaned this up

**Review by halfprice — APPROVED** [review]
LGTM! Thanks for adding all the detailed comments!

I wonder if it also worth creating some kind of cache invariant checker that runs periodically in simtest to catch issues. Some invariants may be that it could compare objects in dirty/cache/db and in object_by_id_cache, to see if they actually match.

**Comment by mystenmark** [comment]
> LGTM! Thanks for adding all the detailed comments!
> 
> I wonder if it also worth creating some kind of cache invariant checker that runs periodically in simtest to catch issues. Some invariants may be that it could compare objects in dirty/cache/db and in object_by_id_cache, to see if they actually match.

adding a debug check

**Comment by halfprice** [comment]
> > LGTM! Thanks for adding all the detailed comments!
> > I wonder if it also worth creating some kind of cache invariant checker that runs periodically in simtest to catch issues. Some invariants may be that it could compare objects in dirty/cache/db and in object_by_id_cache, to see if they actually match.
> 
> adding a debug check

LGTM!

**Pull request merged**